keras_nlp/tests/doc_tests/fenced_docstring_lib.py

```diff
@@ -0,0 +1,208 @@
+          *   in the same binary as any other doctests.
+          *********************************************************************
+          """
+                )
+            )
+            type(self).patched = True
+
+        # Match anything, except if the look-behind sees a closing fence.
+        no_fence = "(.(?<!```))*?"
+        self.fence_cell_re = re.compile(
+            rf"""
+        ^(                             # After a newline
+            \s*```\s*({fence_label})\n   # Open a labeled ``` fence
+            (?P<doctest>{no_fence})      # Match anything except a closing fence
+            \n\s*```\s*(\n|$)            # Close the fence.
+        )
+        (                              # Optional!
+            [\s\n]*                      # Any number of blank lines.
+            ```\s*\n                     # Open ```
+            (?P<output>{no_fence})       # Anything except a closing fence
+            \n\s*```                     # Close the fence.
+        )?
+        """,
+            # Multiline so ^ matches after a newline
+            re.MULTILINE |
+            # Dotall so `.` matches newlines.
+            re.DOTALL |
+            # Verbose to allow comments/ignore-whitespace.
+            re.VERBOSE,
+        )
+
+    def get_examples(
+        self, string: str, name: str = "<string>"
+    ) -> Iterable[doctest.Example]:
+        # Check for a file-level skip comment.
+        if re.search(
+            "<!--.*?doctest.*?skip.*?all.*?-->", string, re.IGNORECASE
+        ):
+            return
+
+        for match in self.fence_cell_re.finditer(string):
+            if re.search("doctest.*skip", match.group(0), re.IGNORECASE):
+                continue
+
+            groups = match.groupdict()
+
+            source = textwrap.dedent(groups["doctest"])
+            want = groups["output"]
+            if want is not None:
+                want = textwrap.dedent(want)
+
+            yield doctest.Example(
+                lineno=string[: match.start()].count("\n") + 1,
+                source=source,
+                want=want,
+            )
+
+
+def _print_if_not_none(obj):
+    """Print like a notebook: Show the repr if the object is not None.
+
+    `_patch_compile` Uses this on the final expression in each cell.
+
+    This way the outputs feel like notebooks.
+
+    Args:
+      obj: the object to print.
+    """
+    if obj is not None:
+        print(repr(obj))
+
+
+def _patch_compile(
+    source, filename, mode, flags=0, dont_inherit=False, optimize=-1
+):
+    """Patch `doctest.compile` to make doctest to behave like a notebook.
+
+    Default settings for doctest are configured to run like a repl: one statement
+    at a time. The doctest source uses `compile(..., mode="single")`
+
+    So to let doctest act like a notebook:
+
+    1. We need `mode="exec"` (easy)
+    2. We need the last expression to be printed (harder).
```

Comment:
I wonder if we want this... Nothing relies on output after a fenced code block today, but we could add this in.

Will discuss with folks.